### DISCUSSION TITLE
Questions about InterfaceKernels/BCs System- implementing custom interface conditions

### DISCUSSION THREAD

**Discussion by richmondodufisan** [idaholab/moose#25746] [Q&A General] [answered]
Hi, I want to understand how the interface kernel system works, as I have custom normal flux conditions as well as tangential flux conditions to implement.

**UNDERSTANDING InterfaceKernels System using SideSetHeatTransferKernel as an example:**

For an interface as described in the documentation (https://mooseframework.inl.gov/source/interfacekernels/SideSetHeatTransferKernel.html), the heat flux on either side of the interface should be equivalent. Considering only conduction for now, we have:

![moose_question](https://github.com/idaholab/moose/assets/52802063/39e09db2-f08b-4b4b-8309-ab1ba9eb6882)

I looked through the file to see how it was implemented and I was wondering some things:

**1.  How is q defined?** the relationship q = -k∇T is strongly imposed in both heat conduction kernels. That is, this expression is directly substituted in the ADHeatConduction kernel, basically eliminating q from the Heat Equation (writing it all in terms of T). So, how does MOOSE know that q refers to the gradient of temperature _multiplied by the thermal conductivity_? Even in the BC system, if we prescribe q for a Neumann BC, _how does it know that the value defined needs to be scaled by the thermal conductivity at that point to relate it to our temperature variable_? Because as far as MOOSE is concerned, q itself doesn't really exist, the both variables and Heat Equation are written entirely in terms of Temperature.

Asking because in the custom formulation I want to implement, q is not strongly connected to Temperature by a simple gradient, it is weakly enforced through kernels that include both q and T (i.e the "Heat Equation" contains both q and T terms i.e q is its own variable), unlike here where q was completely eliminated by a direct substitution. 

**2. Where is q in the interface condition implementation?** Given the definition of the condition (conduction alone), we should have that q_var + C*(T_var - T_neighbor) = 0; and q_neighbor + C*(T_var - T_neighbor) = 0. Looking through the SideSetHeatTransferKernel Source file:

```
  if (_cond[_qp] != 0.0) // Conduction
  {
    Real jump = _u[_qp] - _neighbor_value[_qp];
    switch (type)
    {
      case Moose::Element:
        r += _cond[_qp] * jump * _test[_i][_qp];
        break;

      case Moose::Neighbor:
        r -= _cond[_qp] * jump * _test_neighbor[_i][_qp];
        break;
    }
  }
```

In the above, `jump =_u[_qp] - _neighbor_value[_qp]` is equivalent to (T_var - T_neighbor). I also see that the FE discretization was done twice, for the two cases (i.e _test and _test_neighbor being the test functions multiplied for each case). The only thing that seems to be "missing" is q for each case. Is it automatically assumed that q is added to this for InterfaceKernels? And if so, that brings me back to question 1- how is q defined? As far as MOOSE is concerned, q does not exist, as all the kernels are written COMPLETELY in terms of temperature.

**3. What component of q is this being implemented for?** In the BC system, it is the normal component of the flux. I looked through the NeumannBC source file and found:

```
{
  return -_test[_i][_qp] * _value;
}
```

So my guess is that with `[_i]` its somehow applying it to the normal component. I couldn't find elsewhere in the file where _i was defined, though.

I ask because I want to implement different interface conditions for the NORMAL component and TANGENTIAL component. So, in my interface condition file, I need to be able to specify that it applies only to the specific components of q I need. Basically, _how can I define unit tangent and normal vectors at a boundary so I can multiply my q vector by them?_

**Comment by lindsayad** [comment]
Just considering heat conduction, when you integrate your weak form by parts, you end up with a surface term:

<test, n * -k * grad T>

where n is the surface normal. If an interface kernel implements the residual:

<test, conductance * (T - T_neighbor)>

then it is weakly setting the condition:

n * -k * grad T = conductance * (T - T_neighbor)

**Reply by richmondodufisan** [reply]
I see. So it is able to find out that info based on what I define in the kernels? That, and I assume the BC system does the same thing, i.e uses <test, n * -k * grad T> to apply the Neumann BCs?

Then how could a case where **the discontinuities are not related by the flux** be implemented? For this case, what if , instead of `q_var + C*(T_var - T_neighbor) = 0`, you had `C*(T_var - T_neighbor) + CONSTANT = 0`? 

I'm asking because in my case, q and T are both independent variables weakly related through kernels I implemented. So for my interface condition, I would like to use Coupleable to couple the **values** of q and T at both "var" and "neighbor var", and then implement my custom interface condition.

Also, is there a way to extract the unit normal at a surface in MOOSE? I would need to multiply q by the unit normal. With the unit normal I should be able to find a unit tangent as well.

**Reply by lindsayad** [reply]
> So it is able to find out that info based on what I define in the kernels? That, and I assume the BC system does the same thing, i.e uses <test, n * -k * grad T> to apply the Neumann BCs?

The code isn't finding out anything. It's just how the numerics of the finite element scheme workout. You have to write out your weak form and identify what terms have been integrated by parts. Often it is only the diffusion or conduction portion of the equation, in which case your surface term is <test, n * -k * grad T>. If you don't apply any boundary condition or interface kernel, then you are weakly setting n * -k * grad T = 0. If you write one BC with residual <test, A>, another boundary condition with residual <test, B>, then you will be weakly setting n * -k * grad T = A + B

> Then how could a case where the discontinuities are not related by the flux be implemented? For this case, what if , instead of q_var + C*(T_var - T_neighbor) = 0, you had C*(T_var - T_neighbor) + CONSTANT = 0?

Hopefully you have at least some term that is integrated by parts? You could implement this condition using a penalty method (or in theory a Lagrange multiplier but that takes some more work). You could write a residual that is <test, penalty * (C * (T - Tn) - CONSTANT)>. The larger the penalty factor the more closely you would require your condition to be met. Note that this will still weakly enforce a condition: flux = penalty * (C * (T - Tn) - CONSTANT)

> I'm asking because in my case, q and T are both independent variables weakly related through kernels I implemented. So for my interface condition, I would like to use Coupleable to couple the values of q and T at both "var" and "neighbor var", and then implement my custom interface condition.

If my answers to the above questions don't help, then we can talk more about this.

> Also, is there a way to extract the unit normal at a surface in MOOSE?

Yes, it is `_normals[_qp]`

**Reply by richmondodufisan** [reply]
**********************************************************************************************
Just to be sure I understand how it works correctly, writing the weak form of the heat equation (ignoring velocity effects and assuming no volumetric heat source):


![heat_eqn](https://github.com/idaholab/moose/assets/52802063/e973912b-b122-4581-86f9-6105248fa7b1)

Looking through the source files, the time dependent term is implemented with:

```
ADReal
ADHeatConductionTimeDerivative::precomputeQpResidual()
{
  return _specific_heat[_qp] * _density[_qp] * ADTimeDerivative::precomputeQpResidual();
}
```

The Kernel part is implemented as:

```
ADRealVectorValue
ADHeatConduction::precomputeQpResidual()
{
  return _thermal_conductivity[_qp] * ADDiffusion::precomputeQpResidual();
}
```

Where ADDiffusion, which it inherits from is 

```
ADRealVectorValue
ADDiffusion::precomputeQpResidual()
{
  return _grad_u[_qp];
}
```

And the use of `precomputeQpResidual()` ensures that each one is multiplied by the test function/gradient of test function, depending on which is appropriate.

The BC term is not implemented in any of these. What I meant by "finding out" is, is MOOSE able to automatically know what the BC term looks like based on the kernel term, since they both come from a term that was integrated by parts? That is, does MOOSE work backwards based on what you provide as the kernel to get what the BC term should be like? In the case where nothing was integrated by parts (and hence, no grad_test kernel term), how does MOOSE know what the Neumann condition should be like? Will it just be a derivative of the u term?

**********************************************************************************************

As for the implementation of the custom interface condition, if I use a kernel with a penalty, since it involves two variables that are defined only on separate blocks, will MOOSE only try to implement it at the interface where both variables are defined? I don't want it to try to force that condition within the bulk of the two blocks, only at the interface. But I've seen that most Kernels do not have the "boundary" option, only the "block" option. My interface condition is of the form:

q.n = f(T, Tn, q)

I was also thinking about strongly enforcing the condition by creating a material file where I use coupleable to couple the needed variables, do the calculations, and then output a material that I can apply as a dirichlet condition for each q variable on either side of the boundary. The things I was wondering are:

-This is an implicit expression for q, will it be able to handle that? i.e a BC that basically depends on the variable itself? Getting it explicitly as a function of q would be hard, but easier to do explicitly as a function of each of the T's.

-I saw in the documentation that I can define "boundary" for material files. Will that ensure that the custom values (that become the dirichlet condition) are only calculated on that boundary?

-Since the condition only applies to the normal component of my q variables, I will need to define an aux variable which is the projection of q on the unit normal to do that. I was thinking of outputting the unit normal at each qp as a material, and then doing the dot product calculation in the "expression" field for the AuxKernel. Can BCs be defined for Aux Variables?

-Alternatively, if I define it explicitly as a function of T then I guess I can do the unit normal dot product within the material file.

-For any of this to work, can I define boundary conditions that depend on a variable? That is, the variables coupled within my material file used to calculate the BC.

**********************************************************************************************

**Reply by lindsayad** [reply]
> What I meant by "finding out" is, is MOOSE able to automatically know what the BC term looks like based on the kernel term, since they both come from a term that was integrated by parts? 

I wouldn't say MOOSE automatically knows; I would say "the math" automatically knows. There is nothing in the code that draws a relationship between kernels and boundary conditions.

> In the case where nothing was integrated by parts (and hence, no grad_test kernel term), how does MOOSE know what the Neumann condition should be like?

If no term is integrated by parts, then it does not make sense mathematically to impose an `IntegratedBC` or an `InterfaceKernel`. You would be imposing a boundary or interface flux that should not exist

> As for the implementation of the custom interface condition, if I use a kernel with a penalty, since it involves two variables that are defined only on separate blocks, will MOOSE only try to implement it at the interface where both variables are defined? I don't want it to try to force that condition within the bulk of the two blocks, only at the interface.

Interface kernels only run on the boundary you assign them. Unlike "regular" kernels they are boundary restricted as opposed to block restricted.

> I was also thinking about strongly enforcing the condition by creating a material file where I use coupleable to couple the needed variables, do the calculations, and then output a material that I can apply as a dirichlet condition for each q variable on either side of the boundary.

Imposing Dirichlet conditions at interfaces degrades the accuracy of the finite element method. I would avoid that if possible.

> This is an implicit expression for q, will it be able to handle that? i.e a BC that basically depends on the variable itself? Getting it explicitly as a function of q would be hard, but easier to do explicitly as a function of each of the T's.

All residual objects in MOOSE whether Kernels, IntegratedBCs, NodalBCs (DirichletBCs), or InterfaceKernels fully support implicit treatments. This is the "bread-and-butter" of MOOSE

> I saw in the documentation that I can define "boundary" for material files. Will that ensure that the custom values (that become the dirichlet condition) are only calculated on that boundary?

Yes, those material calculations will be restricted to that boundary.

> Since the condition only applies to the normal component of my q variables, I will need to define an aux variable which is the projection of q on the unit normal to do that. I was thinking of outputting the unit normal at each qp as a material, and then doing the dot product calculation in the "expression" field for the AuxKernel. Can BCs be defined for Aux Variables?

This is becoming quite the chain. If you are using an IntegratedBC or InterfaceKernel you could use the `_normals` member that already exists. As a NodalBC this would be pretty difficult. You cannot use a material property in an aux kernel that runs on nodes. Similarly you could not use an elemental variable or a material property directly in a NodalBC.

> For any of this to work, can I define boundary conditions that depend on a variable? That is, the variables coupled within my material file used to calculate the BC.

You can use any kind of variable in an IntegratedBC or InterfaceKernel. You can only use C0 family variables (e.g. Lagrange) in NodalBCs

**Reply by richmondodufisan** [reply]
This is helpful thanks! I'm going to try to implement first and see how that goes.

**Reply by richmondodufisan** [reply]
Just wanted to provide an update on this, I added a comment to https://github.com/idaholab/moose/discussions/25894 regarding an issue faced with my attempt, i.e using the method of implementing the interface condition as dirichlet BCs.

![formulation](https://github.com/idaholab/moose/assets/52802063/85bd1c64-aa87-4a66-b0ff-07579b2f7c91)

As for using the InterfaceCondition or IntegratedBC, I'm not sure how that would work here noting:

> Note that this will still weakly enforce a condition: flux = penalty * (C * (T - Tn) - CONSTANT)

and I don't have anything to do with the flux here, as I am implementing each T and each component of each q as independent variables (i.e taking this to 8 equations instead of 4).

Also, I am taking the variation of A3b and A1b wrt q, but doing that for A1b means there is no flux term.

**Reply by richmondodufisan** [reply]
Actually, since the integrated BCs 

> weakly enforce a condition: flux = penalty * (C * (T - Tn) - CONSTANT)

would it make sense to provide flux = flux + penalty * (C * (T - Tn) - Expression)

i.e basically hardcoding the BC term there so that as it forces it to equate to itself, the expression has to go to zero? If so I can try to redo the whole thing and take the variation of A1a and A3a wrt to q so I can have flux terms for each. Only issue is that I'll have to deal with the second gradients in A3b but I think MOOSE has a coupledSecond.

**Reply by lindsayad** [reply]
So it seems to me you are using a mixed formulation, with both the temperature and flux as nonlinear variables. Have you looked into Raviart-Thomas finite elements? @nmnobre how are we doing on getting RT into MOOSE? Is it ready?

**Reply by richmondodufisan** [reply]
Yes exactly that. I am not sure how else I would implement otherwise as getting A3b explicitly in terms of q seems like an impossible task, not to mention the interface normal condition (A6).

I have not heard of Raviart-Thomas finite elements before, how would they work in this case?

**Reply by nmnobre** [reply]
_Almost_. I was refactoring last week, and I _think_ I left it at a stage where it compiles and it should work unless I messed up any signs in the kernels, https://github.com/farscape-project/moose/tree/rt. I can have a look on Monday. It's still missing documentation too. The plan was to get those TET14s merged beforehand and since that (and RT) both depend on an up-to-date libMesh I slacked a bit and worked on other stuff. :/

In other words, it's probably in a state where it can be used, not in a state where it can be merged.

**Reply by lindsayad** [reply]
> I have not heard of Raviart-Thomas finite elements before, how would they work in this case?

@nmnobre I actually haven't checked, do we support k > 0 for RT? Additionally, if you had k >= 1, can you integrate the (psi, div(q)) term by parts and move the del operator to the scalar test function (giving you a surface term with n * q) and maintain the proper error convergence? If the answers to both these questions are yes ... I would go with RT, and integrate (psi, div(q)) by parts which would naturally give you a surface term with the flux for which you can implement InterfaceKernels, IntegratedBCs, etc. Otherwise we're stuck with penalty methods (or something else like Lagrange multipliers on the surface or strong constraint enforcement for the normal component of q)

**Reply by richmondodufisan** [reply]
Hmm I think there are some gaps in my knowledge I need to address, its clear I need to have a more fundamental understanding of Mixed Formulations but thank you for your help so far!! I'll be back to this when I make some more progress in that regard.

**Reply by nmnobre** [reply]
> do we support k > 0 for RT?

@lindsayad No. We're currently in the process of deciding if that's something we want to do from early next year. As a matter of fact, your opinion would be most welcome.

>  if you had k >= 1, can you integrate the (psi, div(q)) term by parts and move the del operator to the scalar test function (giving you a surface term with n * q) and maintain the proper error convergence? 

Well, if we do this, we are no longer taking any divergences, right? We'd have (grad psi, q) for the non-boundary term in both equations. At that point, are we, not quite, but basically talking about a classical problem with Neumann BCs on T?

**Reply by lindsayad** [reply]
You would still have the divergence of the test function in the "vector equation". But yea I don't know if you remove the application of the divergence to the vector solution approximation in the "scalar equation" whether it's no longer appropriate to apply H(div) spaces. It sounds like you're telling me it's not? :smile:

**Reply by nmnobre** [reply]
> You would still have the divergence of the test function in the "vector equation".

Ah, in my mind I was applying integration by parts to both equations, so the Jacobian remains symmetric. The divergence of the test function in the vector equation appears by the same integration by parts you are now applying to the scalar equation. I suspect that might be important to match polynomial orders between coupled variables.

**Reply by lindsayad** [reply]
Returning to the problem at hand, my experience with mixed formulations is about a month old. You could definitely do this with a discontinuous Galerkin method. However, I also suspect you can do this with Taylor-Hood finite elements as well. @roystgnr maybe you can weigh in here...

**Reply by nmnobre** [reply]
> Returning to the problem at hand, my experience with mixed formulations is about a month old.

Same here.

**Reply by roystgnr** [reply]
> However, I also suspect you can do this with Taylor-Hood finite elements as well. @roystgnr maybe you can weigh in here...

I think of Taylor-Hood in the context of divergence-free flow problems (that being where everybody's happy about them satisfying the LBB condition without need for stabilization), but we're still just talking about a conduction problem here, right?  So we're thinking of Taylor-Hood in the sense of a mixed method where the order p+1 vector space is q and the order p space is T (with test function psi)?  The standard mixed method there looks like ... (-∇·q,v)=(f,v) and (q,τ) = (-u,∇·(kτ)), plus a boundary term that's a *weak* application of the Dirichlet BC for u, on a space restricted by a *strong* application of the Neumann BC to the normal components of q?

(if it's not yet obvious, despite doing a PhD while surrounded by people doing mixed methods for Darcy flow, I stuck like glue to single-variable H1 formulations of convection-diffusion...)

But then trying to do a strong restriction of just the normal components of q is tricky, so we're hoping to get that back to a weak term too, by integrating (-∇·q,v) to get (q,∇v) along with a boundary (v,q·n)?  You do get a classic [A B; Bᵀ 0] saddle point system there, which screams "LBB condition or need for stabilization" to me, but I have no idea how the condition cashes out.  This looks a little like incompressible Stokes, but with an A block that's more like an L2 norm than an H1 seminorm, which makes me think it'd be more stable and Taylor-Hood would be fine, but that's a hunch, not a proof.

**Reply by lindsayad** [reply]
> But then trying to do a strong restriction of just the normal components of q is tricky, so we're hoping to get that back to a weak term too, by integrating (-∇·q,v) to get (q,∇v) along with a boundary (v,q·n)? You do get a classic [A B; Bᵀ 0] saddle point system there, which screams "LBB condition or need for stabilization" to me

Yes, these were all my thoughts as well

**Reply by lindsayad** [reply]
So @richmondodufisan I would try a formulation with a `SECOND LAGRANGE_VEC` for q, and a `FIRST LAGRANGE` for T

**Reply by lindsayad** [reply]
And I would try integrating the second term on the LHS of equation A1a by parts

**Reply by richmondodufisan** [reply]
I'm going to be honest, I got a little bit lost reading through the replies and wasn't able to fully follow😅 but if I understand your suggestion correctly, I can try to

-Use second order lagrange shape functions for q

-Take the variation of A1a and A3a wrt to q, that way I have a boundary term in q

-Use first order shape functions for T

-Take the variation of A1b and A3b wrt to T, and use coupledSecond and whatever other coupled methods I need to evaluate the second order derivatives of q in A3b

-With this setup, I now have boundary terms of the form test*q.n

-I can then rearange A6 to be in the form q.n = normal_condition, and then implement an interface kernel with residual = normal_condition * test

-I do have a slip condition for the tangential component of q as well though (below), would it make sense to bring all terms to one side to equal zero, and then tack on a penalty, so that the interface residual now becomes residual = (normal condition + penalty * tangent_condition) * test for one side, and residual = normal condition*test for the other, effectively forcing the tangent condition to be zero?

![a7](https://github.com/idaholab/moose/assets/52802063/b3beffc1-5665-45fe-a3b1-e535aacba725)

Is this what you're suggesting?

**Reply by lindsayad** [reply, marked as answer]
Don't change the variation application. You have to take the variation of the vector valued equation with respect to q and you have to take the variation of the scalar valued equation with respect to T. I think the only change I'm asking for with respect to your weak form is that you should integrate the term (test, div(q)), which comes from the second term of A1a, by parts. If you do that, you will wind up with two terms:

<test, n * q> - (grad test, q)

where here "test" is the scalar test function, angle brackets denote boundary integration, and parentheses denote volume integration

**Reply by lindsayad** [reply]
The tangential condition really does muddy things up

**Reply by lindsayad** [reply]
For flow modeling, a slip condition is often imposed by simply setting the normal component of the velocity to zero and just letting the tangential velocity "be"

**Reply by lindsayad** [reply]
At this point I've run out of good ideas for you. I would be wary of doing this:

> so that the interface residual now becomes residual = (normal condition + penalty * tangent_condition) * test for one side, and residual = normal condition*test for the other

because if the tangential portion is not 0 then this is not conservative, e.g. you have a different amount of stuff leaving one side than entering another

**Reply by richmondodufisan** [reply]
Ah yeah that makes sense. So then the interface condition will be expressed for the boundary term in T which includes q. That makes sense, I'll try that.

As for the tangent condition I think I'll have to make some assumptions to get rid of it but everything else should work as described, thank you! I'll give the implementation a go.

As a side note, while searching I did see in the documentation plans to include a refined PenaltyDirichletBC based on https://www.ams.org/journals/mcom/2009-78-267/S0025-5718-08-02183-2/home.html, would love to know if that's still in the works or not.

**Reply by lindsayad** [reply]
I don't think we've done that. Maybe @hugary1995 can weigh in. He knows stuff about Nitsche

**Reply by hugary1995** [reply]
@ttruster has an app for various nitsche-type enforcements. I don't remember if it's in-house or open-source.

**Reply by hugary1995** [reply]
There are some interface condition in the xfem module based on the nitsche formulation. Though the last time I checked they were derived based on the diffusion problem. I didn't follow this entire thread, so I am not sure how useful it will be to you at the moment.

**Reply by lindsayad** [reply]
It's a good point that I believe Nitsche conditions have to be derived based on your problem, so if someone has implemented a Nitsche condition for an equation form X, it won't work if you have an equation form Y ... ?

**Reply by hugary1995** [reply]
Yeah, it's more work implementation wise, though it's not all that hard to derive. IIRC, the real challenge for us was that the formulation includes stateful quantities (such as stress in a large kinematics problem), while many of our interfaces are evolving.

**Reply by ttruster** [reply]
Yes our group has been working on Nitsche-type implementations that won't be ready for public release anytime soon. Yes there are extra average terms of gradients from both sides that are needed; but for many weak forms, these terms can be expressed using material properties that can be borrowed from other objects. For example, a unified implementation of small and large strain solid mechanics Nitsche methods is possible.

**Reply by richmondodufisan** [reply]
Hi @lindsayad so I did this and implemented and it wasn't converging, and then I thought to break it down into a smaller problem to find the issue.

So then I implemented just the Fourier heat equation i.e the same one in the HeatConduction Module, this time with q and T as independent variables. No interface condition, nothing fancy. But it doesn't converge. I'm pretty stumped and I've looked at my code multiple times but I can't seem to figure out why, other than maybe I don't have the syntax right. Could you please take a look?

https://github.com/idaholab/moose/discussions/26021
